packages/wordpress-playground-block/src/base64.ts

```diff
@@ -0,0 +1,106 @@
+/**
+ * Base64 encoding and decoding functions.
+ * We cannot just use `btoa` and `atob` because they do not
+ * support Unicode characters.
+ */
+
+export const attributesToBase64 = [
+	'blueprint',
+	'blueprintUrl',
+	'codeEditorErrorLog',
+	'constants',
+	'files',
+];
+
+export function base64EncodeBlockAttributes(
+	blockAttributes: Record<string, any>
+) {
+	const base64Props: Record<string, string> = {};
+	for (const key in blockAttributes) {
+		if (
+			!attributesToBase64.includes(key) ||
+			typeof blockAttributes[key] === 'number' ||
+			typeof blockAttributes[key] === 'boolean' ||
+			typeof blockAttributes[key] === null ||
+			typeof blockAttributes[key] === undefined
+		) {
+			base64Props[key] = blockAttributes[key];
+			continue;
+		}
+		base64Props[key] = stringToBase64(
+			JSON.stringify(blockAttributes[key])
+		);
+	}
+	// The "files" attribute is of type array
+	if ('files' in base64Props) {
+		base64Props['files'] = [base64Props['files']] as any;
+	}
+	return base64Props;
+}
+
+/**
+ * Turns base64 encoded attributes back into their original form.
+ * It never throws, bales out early if we can't decode, and always
+ * returns a valid object. If any attribute cannot be decoded, it
+ * will be kept in its original form and presumed to have a non-base64
+ * value to keep the older version of the block working without
+ * migrating the attributes.
+ *
+ * @param base64Attributes
+ * @returns
+ */
+export function base64DecodeBlockAttributes(
+	base64Attributes: Record<string, any>
+) {
+	const attributes: Record<string, any> = {};
+	for (const key in base64Attributes) {
+		let valueToDecode = base64Attributes[key];
+		// The "files" attribute is of type array
+		if (key === 'files') {
+			valueToDecode = valueToDecode[0];
+		}
+		if (
+			!attributesToBase64.includes(key) ||
+			!(typeof valueToDecode === 'string')
+		) {
+			attributes[key] = base64Attributes[key];
+			continue;
+		}
+		if (key in base64Attributes) {
+			try {
+				attributes[key] = JSON.parse(base64ToString(valueToDecode));
+			} catch (error) {
+				// Ignore errors and keep the base64 encoded string.
+				attributes[key] = base64Attributes[key];
```

Comment:
this looks like it should preserve previous attributes which aren't base-64-encoded, but perhaps accidentally so.

I think it's best if we do that: return the raw string if it doesn't base-64-decode, but it could be nice indicating that here.

Reply:
Done in https://github.com/WordPress/playground-tools/commit/b189b9fa47e830320710e848f186922299c59edf